python/google/protobuf/internal/well_known_types.py

```diff
@@ -372,6 +399,34 @@ def _NormalizeDuration(self, seconds, nanos):
     self.seconds = seconds
     self.nanos = nanos
 
+  def _comparable_scalar(self, obj):
+    if isinstance(obj, timedelta):
+      return (obj.seconds + obj.days * _SECONDS_PER_DAY) * _NANOS_PER_SECOND + obj.microseconds * _NANOS_PER_MICROSECOND
+    return obj.ToNanoseconds()
+
+  def __eq__(self, other):
```

Comment:
These rich comparisons [just won't work in Python2](https://github.com/python/cpython/blob/2.7/Modules/datetimemodule.c#L1757).

Reply:
Isn't the comparisions implemented by yourself, why related to python2/python3? If there's difference for timedelta, you could check the sys.version_info to do different code for python2/python3"
https://github.com/protocolbuffers/protobuf/blob/fd4baa91e7c127a29ee6d5f6ecf0fa149f6d9cec/python/compatibility_tests/v2.5.0/setup.py#L10

Reply:
Unfortunately the C code for python2 has some [oddities](https://github.com/protocolbuffers/protobuf/pull/5168#discussion_r220855631) that stop comparisons from simply working. The fix for `datetime` is simple and not too messy, but for [`timedelta`](https://github.com/python/cpython/blob/2.7/Modules/datetimemodule.c#L1757) I think they can only ever compare to objects that are `timedelta` instances, so the only way may be to subclass from `timedelta`.

Reply:
In that case, maybe we only provide Timestamp compare with Timestamp not datetime, Duration only compare with Duration not timedelta. Users should call ToDatetime/FromDatetime/ToTimedelta/FromTimedelta first if them want to compare with datetime/timedelta

Reply:
It turns out that either `Duration`/`timedelta` comparisons either aren't possible in python2 (tried dropping in `timdelta` as `Duration`'s parent class), or would take involved changes and still leave a mess in the code. I doubt anyone likes the idea of introducing introducing python 2-3 incompatibilities, so that is gone.

`Timestamp`/`datetime` is something I was erring on on, as I think it would be nice to have, and the cost of it is 1 junk attribute, but then it'd be a bit surprising to users that built in type comparisons aren't all implemented, so seems like a fair reason to avoid that junk.

I updated the PR to drop the built in type comparisons, they would be good to add when python2 support is dropped though - is there an issue to drop 2 (even if support will go on for years)?